### DISCUSSION TITLE
Vercel build preset per environment - CRA migration to nextjs

### DISCUSSION THREAD

**Discussion by geositta** [vercel/next.js#38388] [Help]
We have a live project deployed on Vercel in production with CRA that we are migrating to a nextjs app. We are code complete and are able to build the nextjs migrated project locally but not in Vercel. My question is whether or not it is possible to have separate build configurations on Vercel for production and development?

Ideally we would be able to use the CRA template for our production build. But while in development we could use the nextjs build config template for our still in progress nextjs app migration.
<img width="1211" alt="Screen Shot 2022-07-06 at 5 44 38 PM" src="https://user-images.githubusercontent.com/85914066/177655779-8b619b2d-ee40-4f61-a829-a9b361442bb1.png">
<img width="941" alt="Screen Shot 2022-07-06 at 5 45 09 PM" src="https://user-images.githubusercontent.com/85914066/177655782-117cd92f-d308-4283-b020-dc2eb8f5b946.png">

**Comment by geositta** [comment]
I am using the  "Create React App" framework preset because that is what we need in Production. Thus I cannot change that to nextjs preset without affecting our Production builds. Ideally Vercel would give users the ability to set the framework preset by environment.

**Reply by kirstywilliams** [reply]
This is old but for future eyes on this you can add `"framework": "..."` to vercel.json, more info [here](https://vercel.com/docs/concepts/projects/project-configuration#framework)

**Comment by icyJoseph** [comment]
Hi,

I don't know if there's a magic toggle or option in the dashboard to do this, but I reckon you could write a `build.js` script, and run it as `node build.js` on your build step.

The `build.js` step would check an env var (so it might need to use dotenv or something to load env vars, unless these are available to the running environment), and based on `VERCEL_ENV`, as documented here: https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables - please do read through on how to activate these Vercel env vars.

Once your script knows if you are in `vercel` preview or production you can spawn a child process to either execute `npx run react-scripts build` or `npx next run build`.

A good thing is that you could test this locally, just need to provide the VERCEL_ENV locally.

```js
const { exec } = require("child_process");

const isProd = process.env.FOO === "bar";

// instead of ls -l, use the react-scripts build command npx react-scripts build (?)
const command = isProd ? "ls -l" : "npx next build";

const child = exec(command, function (error, stdout, stderr) {
  if (error) {
    console.log(error.stack);
    console.log("Error code: " + error.code);
    console.log("Signal received: " + error.signal);
  }
  console.log(stdout);
  console.warn(stderr);
});

child.on("exit", function (code) {
  console.log("Child process exited with exit code " + code);
});
```

If I run the above as `FOO=bar node build.js` then it runs Next.js build, otherwise it just prints out `ls` results.

### Edit

Simplified the script.
